Modules/posixmodule.c

```diff
@@ -5553,6 +5520,103 @@ os__path_normpath_impl(PyObject *module, path_t *path)
     return result;
 }
 
+/*[clinic input]
+os._path_abspath
+
+    path: path_t(make_wide=True, nonstrict=True)
+    /
+
+Return an absolute path.
+[clinic start generated code]*/
+
+static PyObject *
+os__path_abspath_impl(PyObject *module, path_t *path)
+/*[clinic end generated code: output=bb40fbf3be7251a4 input=8870bbcaa7c920a1]*/
+{
+    Py_ssize_t abs_len;
+    wchar_t *abs, *abs_buf = NULL, *cwd_buf = NULL;
+    PyObject *result = NULL;
+
+    wchar_t *path_buf = (wchar_t *)path->wide;
+    Py_ssize_t path_len = path->length;
+    int use_bytes = PyBytes_Check(path->object);
+#ifdef MS_WINDOWS
+    if (wcslen(path_buf) != path_len) {
+        PyErr_Format(PyExc_ValueError,
+                     "_path_abspath: embedded null character in path");
+        goto exit;
+    }
+    // Preserve `.\` for qualified referencing
+    abs = _Py_normpath_and_size(path_buf, path_len, 0, &abs_len, 1);
+    if (abs_len == 0 || (abs_len == 1 && abs[0] == L'.')) {
+        result = posix_getcwd(use_bytes);
+        goto exit;
+    }
+    if (_PyOS_getfullpathname(abs, &abs_buf) < 0) {
+        result = win32_error_object("GetFullPathNameW", path->object);
+        goto exit;
+    }
+    abs = abs_buf;
+    abs_len = wcslen(abs_buf);
+#else
+    if (path_len == 0 || (path_len == 1 && path_buf[0] == L'.')) {
+        result = posix_getcwd(use_bytes);
+        goto exit;
+    }
+
+    if (_Py_isabs(path_buf)) {
+        abs = _Py_normpath_and_size(path_buf, path_len, 0, &abs_len, 0);
+    }
+    else {
+        PyObject *cwd_obj = posix_getcwd(0);
+        if (!cwd_obj) {
+            goto exit;
+        }
+        Py_ssize_t cwd_len;
+        cwd_buf = PyUnicode_AsWideCharString(cwd_obj, &cwd_len);
+        Py_DECREF(cwd_obj);
+        if (!cwd_buf) {
+            goto exit;
+        }
+
+        int add_sep = cwd_buf[cwd_len - 1] != SEP;
+        Py_ssize_t prefix_len = cwd_len + add_sep;
+        abs_len = prefix_len + path_len;
+
+        if ((size_t)abs_len + 1 > (size_t)PY_SSIZE_T_MAX / sizeof(wchar_t)) {
+            PyErr_SetString(PyExc_OverflowError, "path is too long");
+            goto exit;
+        }
+
+        abs_buf = PyMem_RawMalloc(((size_t)abs_len + 1) * sizeof(wchar_t));
+        if (!abs_buf) {
+            result = PyErr_NoMemory();
```

Comment:
`result` is initialized to `NULL`; no need to reset it.

```suggestion
            PyErr_NoMemory();
```